### Beginning of the Assignment - exploration

In [6]:
import pandas as pd
import numpy as np
import seaborn as sns
#pd.set_option('max_columns', 200)

### PHIL DATA SET

In [7]:
character_nicknames_df = pd.read_csv('datasets/PHIL DATA SET/character_nicknames.csv')

In [10]:
details_df = pd.read_csv('datasets/PHIL DATA SET/details.csv')

In [11]:
favs_df = pd.read_csv('datasets/PHIL DATA SET/favs.csv')


In [12]:
person_details_df = pd.read_csv('datasets/PHIL DATA SET/person_details.csv')


In [13]:
person_alternate_names_df = pd.read_csv('datasets/PHIL DATA SET/person_alternate_names.csv')


In [14]:
person_anime_works_df = pd.read_csv('datasets/PHIL DATA SET/person_anime_works.csv')


In [15]:
stats_df = pd.read_csv('datasets/PHIL DATA SET/stats.csv')


### DENIS DATA SET

In [16]:
ratings_df = pd.read_csv('datasets/DENIS DATA SET/ratings.csv')

In [8]:
characters_df = pd.read_csv('datasets/DENIS DATA SET/characters.csv')

In [9]:
character_anime_works_df = pd.read_csv('datasets/DENIS DATA SET/character_anime_works.csv')

In [10]:
person_voice_works_df = pd.read_csv('datasets/DENIS DATA SET/person_voice_works.csv')

In [18]:
profiles_df = pd.read_csv('datasets/DENIS DATA SET/profiles.csv')

In [12]:
recommendations_df = pd.read_csv('datasets/DENIS DATA SET/recommendations.csv')

##### First look

In [22]:
character_nicknames_df.shape

(37080, 2)

In [23]:
character_nicknames_df.head()

character_mal_id              nickname
0            280205              Hikaruko
1            280129              Hinacchi
2            280127         Bertha Willis
3            280066                 Jimmy
4            280059  Full Body Red Square

In [24]:
character_nicknames_df.columns
#will list all the columns. Not necessary here but kept as a pattern to follow with the following files

Index(['character_mal_id', 'nickname'], dtype='object')

In [25]:
character_nicknames_df.dtypes

character_mal_id     int64
nickname            object
dtype: object

In [26]:
character_nicknames_df.describe()
#not needed here because it should just be ids

character_mal_id
count      37080.000000
mean      115767.769579
std        87596.063426
min            3.000000
25%        30541.750000
50%       109656.000000
75%       192520.750000
max       282255.000000

### Data preparation (cleaning)


##### On the first dataset we may need to check for duplicates.
What does that mean? We have 102 rows that are duplicated over a 37080 rows dataset.
Why is that? Are there some characters that have multiple nicknames, so they are repeated in the dataset?

In [27]:
character_nicknames_df.loc[character_nicknames_df.duplicated()]
#by default will give us the second

character_mal_id        nickname
328              276501            Koko
329              276501         Cao Cao
352              275883            King
356              275835  Eldest Brother
447              274437   The Half-Fool
...                 ...             ...
37051            281206          Apemon
37052            281206          Apemon
37053            281206          Apemon
37054            281206          Apemon
37068            281717         Colosso

[102 rows x 2 columns]

In [28]:
character_nicknames_df.query('character_mal_id == 75628')

character_mal_id        nickname
21558             75628        Anneroze
21559             75628  Witch of Steel
21560             75628      Iron Witch
21561             75628        Anneroze
21562             75628  Witch of Steel
21563             75628      Iron Witch

Mhh they're somehow different so yeah, the same character could have different nicknames.
We want to drop though the ones that are exactly the same.

In [29]:
#this way we drop the duplicates on the first dataset

character_nicknames_df = character_nicknames_df\
    .loc[~character_nicknames_df.duplicated()]\
    .reset_index(drop=True).copy()

#we don't need to use a subset here because there are just 2 columns

Let's check for nan values

In [30]:
character_nicknames_df[character_nicknames_df.isna().any(axis=1)]


character_mal_id nickname
687              271183      NaN
3132             248157      NaN
4842             230372      NaN
15847            135548      NaN
18544            107631      NaN
24213             48077      NaN
24976             42325      NaN
26539             35444      NaN
30050             19532      NaN
30051             19531      NaN
30053             19527      NaN
30149             19265      NaN
30225             18984      NaN
30235             18952      NaN
31251             15681      NaN
33882              6562      NaN

In [31]:
#cleaning the df from nan values
character_nicknames_df = character_nicknames_df.dropna()


## SECOND DATASET

##### On the second dataset we may need to check for missing values and/or inconsistent values, since there are no duplicates

In [32]:
details_df.isna().sum()

mal_id                 0
title                  0
title_japanese       119
url                    0
image_url              0
type                  67
status                 0
score              10073
scored_by          10073
start_date           851
end_date           17688
synopsis            5127
rank                6958
popularity             0
members                0
favorites              0
genres                 0
studios                0
themes                 0
demographics           0
source                 0
rating               600
episodes             680
season             22689
year               22689
producers              0
explicit_genres        0
licensors              0
streaming              0
dtype: int64

In [33]:
details_df
#anime details
#japanes title could be dropped?
#members stand for how many users have this anime added to their list.
#explicit_genres is empty so can be removed
#licensor and streaming are mostly empty. Do we care?

mal_id                                              title  \
0       59356                                           -Socket-   
1       56036                                             ......   
2        2928                               .hack//G.U. Returner   
3        3269                                .hack//G.U. Trilogy   
4        4469                   .hack//G.U. Trilogy: Parody Mode   
...       ...                                                ...   
28950   59421  Zutaboro Reijou wa Ane no Moto Konyakusha ni D...   
28951   31245  Zutto Mae kara Suki deshita. Kokuhaku Jikkou I...   
28952   36305  Zutto Mae kara Suki deshita. Kokuhaku Jikkou I...   
28953   34895                                   Zutto Suki Datta   
28954   56606                                    Zutto Tomodachi   

                         title_japanese  \
0                              -socket-   
1                                ......   
2                  .HACK//G.U. RETURNER   
3                   .hack//G.U. Trilogy   
4                   .hack//G.U. Trilogy   
...                                 ...   
28950               ずたぼろ令嬢は姉の元婚約者に溺愛される   
28951             ずっと前から好きでした。～告白実行委員会～   
28952  ずっと前から好きでした。～告白実行委員会～ 「金曜日のおはよう」   
28953                          ずっと好きだった   
28954                             ずっと友達   

                                                     url  \
0           https://myanimelist.net/anime/59356/-Socket-   
1                  https://myanimelist.net/anime/56036/-   
2      https://myanimelist.net/anime/2928/hack__GU_Re...   
3      https://myanimelist.net/anime/3269/hack__GU_Tr...   
4      https://myanimelist.net/anime/4469/hack__GU_Tr...   
...                                                  ...   
28950  https://myanimelist.net/anime/59421/Zutaboro_R...   
28951  https://myanimelist.net/anime/31245/Zutto_Mae_...   
28952  https://myanimelist.net/anime/36305/Zutto_Mae_...   
28953  https://myanimelist.net/anime/34895/Zutto_Suki...   
28954  https://myanimelist.net/anime/56606/Zutto_Tomo...   

                                               image_url     type  \
0      https://cdn.myanimelist.net/images/anime/1043/...    Movie   
1      https://cdn.myanimelist.net/images/anime/1057/...    Music   
2      https://cdn.myanimelist.net/images/anime/1798/...      OVA   
3      https://cdn.myanimelist.net/images/anime/1566/...    Movie   
4      https://cdn.myanimelist.net/images/anime/10/86...  Special   
...                                                  ...      ...   
28950  https://cdn.myanimelist.net/images/anime/1518/...       TV   
28951  https://cdn.myanimelist.net/images/anime/3/821...    Movie   
28952  https://cdn.myanimelist.net/images/anime/6/883...  Special   
28953  https://cdn.myanimelist.net/images/anime/1498/...      OVA   
28954  https://cdn.myanimelist.net/images/anime/1462/...    Music   

                status  score  scored_by                 start_date  ...  \
0      Finished Airing    NaN        NaN  2010-01-01T00:00:00+00:00  ...   
1      Finished Airing   6.53      503.0  2023-06-11T00:00:00+00:00  ...   
2      Finished Airing   6.65     9745.0  2007-01-18T00:00:00+00:00  ...   
3      Finished Airing   7.06    15373.0  2007-12-22T00:00:00+00:00  ...   
4      Finished Airing   6.35     4317.0  2008-03-25T00:00:00+00:00  ...   
...                ...    ...        ...                        ...  ...   
28950  Finished Airing   7.37    15624.0  2025-07-05T00:00:00+00:00  ...   
28951  Finished Airing   7.20   104106.0  2016-04-23T00:00:00+00:00  ...   
28952  Finished Airing   7.17    10038.0  2016-10-26T00:00:00+00:00  ...   
28953  Finished Airing   5.68     1887.0  2017-04-21T00:00:00+00:00  ...   
28954  Finished Airing    NaN        NaN  1996-06-01T00:00:00+00:00  ...   

      demographics       source                     rating  episodes  season  \
0               []     Original               G - All Ages       1.0     NaN   
1               []     Original  PG-13 - Teens 13 

In [34]:
details_df.describe()

mal_id         score     scored_by          rank    popularity  \
count  28955.000000  18882.000000  1.888200e+04  21997.000000  28955.000000   
mean   33977.521948      6.390500  2.996308e+04  11033.157794  14500.358798   
std    19616.858566      0.892045  1.219666e+05   6388.279473   8373.663464   
min        1.000000      1.890000  1.010000e+02      1.000000      1.000000   
25%    15454.000000      5.770000  3.322500e+02   5505.000000   7247.500000   
50%    37969.000000      6.360000  1.528000e+03  11008.000000  14497.000000   
75%    50434.500000      7.030000  1.014575e+04  16515.000000  21754.500000   
max    62590.000000      9.290000  2.979733e+06  22020.000000  28999.000000   

            members      favorites      episodes         year  
count  2.895500e+04   28955.000000  28275.000000  6266.000000  
mean   3.875337e+04     430.848973     14.130080  2010.034153  
std    1.673763e+05    4520.610176     47.161445    13.200708  
min    2.300000e+01       0.000000      1.000000  1961.000000  
25%    2.330000e+02       0.000000      1.000000  2005.000000  
50%    1.077000e+03       1.000000      2.000000  2014.000000  
75%    9.193500e+03      17.000000     13.000000  2020.000000  
max    4.230312e+06  243358.000000   3000.000000  2026.000000

In [35]:
details_df.query('rank == 1')

mal_id              title title_japanese  \
23116   52991  Sousou no Frieren       葬送のフリーレン   

                                                     url  \
23116  https://myanimelist.net/anime/52991/Sousou_no_...   

                                               image_url type  \
23116  https://cdn.myanimelist.net/images/anime/1015/...   TV   

                status  score  scored_by                 start_date  ...  \
23116  Finished Airing   9.29   732845.0  2023-09-29T00:00:00+00:00  ...   

      demographics source                     rating  episodes  season  \
23116  ['Shounen']  Manga  PG-13 - Teens 13 or older      28.0    fall   

         year                                          producers  \
23116  2023.0  ['Aniplex', 'Dentsu', 'Shogakukan-Shueisha Pro...   

      explicit_genres        licensors  \
23116              []  ['Crunchyroll']   

                                               streaming  
23116  ['Crunchyroll', 'Netflix', 'Aniplus TV', 'Baha...  

[1 rows x 29 columns]

In [36]:
details_df[['start_date','season']].query("season.notna()")
#season can be removed? Do we care about the season? We can "calculate" it from the "start_date" field

start_date  season
10     2006-04-06T00:00:00+00:00  spring
11     2002-04-04T00:00:00+00:00  spring
12     2003-01-09T00:00:00+00:00  winter
19     2024-07-07T00:00:00+00:00  summer
20     2025-01-05T00:00:00+00:00  winter
...                          ...     ...
28906  2001-04-07T00:00:00+00:00  spring
28913  2011-05-18T00:00:00+00:00  spring
28932  1980-04-15T00:00:00+00:00  spring
28939  2012-04-03T00:00:00+00:00  spring
28950  2025-07-05T00:00:00+00:00  summer

[6266 rows x 2 columns]

In [37]:
details_df.query("episodes > 2500")

mal_id    title title_japanese  \
13697    9947  Lan Mao      蓝猫淘气3000问   

                                              url  \
13697  https://myanimelist.net/anime/9947/Lan_Mao   

                                               image_url type  \
13697  https://cdn.myanimelist.net/images/anime/5/275...   TV   

                status  score  scored_by                 start_date  ...  \
13697  Finished Airing   6.06      228.0  1999-10-08T00:00:00+00:00  ...   

      demographics    source         rating  episodes  season  year  \
13697           []  Original  PG - Children    3000.0     NaN   NaN   

                                        producers explicit_genres licensors  \
13697  ['Beijing Sunchime Happy Culture Company']              []        []   

      streaming  
13697        []  

[1 rows x 29 columns]

In [38]:
details_df.dtypes
#scoredby, rank, episodes, year can be an int instead of a float
#start and end dates are not objects but dates
#do we need to swap the empty [] with Nan or not? WE should in order to be able to use the .isna() method and other pandas methods


mal_id               int64
title               object
title_japanese      object
url                 object
image_url           object
type                object
status              object
score              float64
scored_by          float64
start_date          object
end_date            object
synopsis            object
rank               float64
popularity           int64
members              int64
favorites            int64
genres              object
studios             object
themes              object
demographics        object
source              object
rating              object
episodes           float64
season              object
year               float64
producers           object
explicit_genres     object
licensors           object
streaming           object
dtype: object

### THIRD DATASET

In [39]:
favs_df

username   fav_type      id
0          ishikawas      anime   45649
1          ishikawas      anime   38680
2          ishikawas      anime     795
3          ishikawas      anime   37510
4          ishikawas      anime     820
...              ...        ...     ...
4178742  vincent0607  character     497
4178743  vincent0607  character  118739
4178744  vincent0607  character  188177
4178745  vincent0607  character  141354
4178746  vincent0607  character  195178

[4178747 rows x 3 columns]

In [40]:
favs_df.isna().sum()


username    4
fav_type    0
id          0
dtype: int64

In [41]:
favs_df.dtypes

username    object
fav_type    object
id           int64
dtype: object

### FOURTH DATASET

In [42]:
person_alternate_names_df

person_mal_id             alt_name
0                  1         Seki Mondoya
1                  1                 門戸 開
2                  1         Monto Hiraku
3                  3                 雪野五月
4                 10        Kevin Hatcher
...              ...                  ...
20460          89567              Sydsnap
20461          89567  Queen of Degeneracy
20462          89826                   陳浩
20463          89842              Chidori
20464          89842                 Nobu

[20465 rows x 2 columns]

In [43]:
person_alternate_names_df.dtypes

person_mal_id     int64
alt_name         object
dtype: object

In [44]:
person_alternate_names_df.isna().sum()

person_mal_id     0
alt_name         18
dtype: int64

In [45]:
person_alternate_names_df = person_alternate_names_df.dropna()


### FIFTH DATASET

In [46]:
#let's check for duplicates

In [47]:
person_details_df

person_mal_id                                                url  \
0                  1     https://myanimelist.net/people/1/Tomokazu_Seki   
1                  2   https://myanimelist.net/people/2/Tomokazu_Sugita   
2                  3    https://myanimelist.net/people/3/Satsuki_Yukino   
3                  4        https://myanimelist.net/people/4/Aya_Hirano   
4                  5  https://myanimelist.net/people/5/Kenichi_Suzumura   
...              ...                                                ...   
76694          90011  https://myanimelist.net/people/90011/Nanako_Ki...   
76695          90012         https://myanimelist.net/people/90012/Pamon   
76696          90013  https://myanimelist.net/people/90013/Tomoru_Emoto   
76697          90014        https://myanimelist.net/people/90014/Hirari   
76698          90015  https://myanimelist.net/people/90015/Daisuke_T...   

                   website_url  \
0                          NaN   
1          https://agrs.co.jp/   
2                          NaN   
3         http://ayahirano.jp/   
4      https://intention-k.com   
...                        ...   
76694                      NaN   
76695                      NaN   
76696                      NaN   
76697   https://hirari.2-d.jp/   
76698                      NaN   

                                               image_url               name  \
0      https://cdn.myanimelist.net/images/voiceactors...      Tomokazu Seki   
1      https://cdn.myanimelist.net/images/voiceactors...    Tomokazu Sugita   
2      https://cdn.myanimelist.net/images/voiceactors...     Satsuki Yukino   
3      https://cdn.myanimelist.net/images/voiceactors...         Aya Hirano   
4      https://cdn.myanimelist.net/images/voiceactors...   Kenichi Suzumura   
...                                                  ...                ...   
76694  https://cdn.myanimelist.net/img/sp/icon/apple-...   Nanako Kishimoto   
76695  https://cdn.myanimelist.net/img/sp/icon/apple-...              Pamon   
76696  https://cdn.myanimelist.net/img/sp/icon/apple-...       Tomoru Emoto   
76697  https://cdn.myanimelist.net/img/sp/icon/apple-...             Hirari   
76698  https://cdn.myanimelist.net/img/sp/icon/apple-...  Daisuke Taniguchi   

      given_name family_name                   birthday  favorites  \
0             智一           関  1972-09-08T00:00:00+00:00       6219   
1             智和          杉田  1980-10-11T00:00:00+00:00      47666   
2            さつき         ゆきの  1970-05-25T00:00:00+00:00       1777   
3              綾          平野  1987-10-08T00:00:00+00:00      18374   
4             健一          鈴村  1974-09-12T00:00:00+00:00       5176   
...          ...         ...                        ...        ...   
76694         七子          岸本                        NaN          0   
76695        NaN          파몬                        NaN          0   
76696        ともる          柄本                        NaN          0   
76697        NaN         ひらり                        NaN          0   
76698         大介          谷口                        NaN          0   

             relevant_location  
0              Berlin, Germany  
1             Los Angeles, USA  
2                Madrid, Spain  
3                Paris, France  
4                 Osaka, Japan  
...                        ...  
76694            Mumbai, India  
76695             Tokyo, Japan  
76696             Tokyo, Japan  
76697            Paris, France  
76698  Cape Town, South Africa  

[76699 rows x 10 columns]

In [48]:
person_details_df.loc[person_details_df['person_mal_id'].duplicated()]
#we found that the duplicates differ for the "relevant_location" field, which has no interest for us so we drop the duplicates
#TODO

person_mal_id                                                url  \
4171            4564  https://myanimelist.net/people/4564/Natsuko_Ta...   
4172            4564  https://myanimelist.net/people/4564/Natsuko_Ta...   
5202            5742  https://myanimelist.net/people/5742/Sarah_Strange   
6018            6638  https://myanimelist.net/people/6638/Toshio_Suzuki   
6019            6638  https://myanimelist.net/people/6638/Toshio_Suzuki   
...              ...                                                ...   
64387          77565  https://myanimelist.net/people/77565/Ena_Nishi...   
64402          77579   https://myanimelist.net/people/77579/Amina_Gaede   
64412          77588  https://myanimelist.net/people/77588/Willie_Ra...   
64414          77589    https://myanimelist.net/people/77589/Brandi_Ray   
64417          77591  https://myanimelist.net/people/77591/Ilka_Pinh...   

                                  website_url  \
4171                                      NaN   
4172                                      NaN   
5202                                      NaN   
6018                                      NaN   
6019                                      NaN   
...                                       ...   
64387                                     NaN   
64402  https://www.instagram.com/amina.gaede/   
64412                                     NaN   
64414                                     NaN   
64417                                     NaN   

                                               image_url               name  \
4171   https://cdn.myanimelist.net/images/voiceactors...  Natsuko Takahashi   
4172   https://cdn.myanimelist.net/images/voiceactors...  Natsuko Takahashi   
5202   https://cdn.myanimelist.net/images/voiceactors...      Sarah Strange   
6018   https://cdn.myanimelist.net/images/voiceactors...      Toshio Suzuki   
6019   https://cdn.myanimelist.net/images/voiceactors...      Toshio Suzuki   
...                                                  ...                ...   
64387  https://cdn.myanimelist.net/img/sp/icon/apple-...      Ena Nishikawa   
64402  https://cdn.myanimelist.net/images/voiceactors...        Amina Gaede   
64412  https://cdn.myanimelist.net/images/voiceactors...     Willie Ray Jr.   
64414  https://cdn.myanimelist.net/images/voiceactors...         Brandi Ray   
64417  https://cdn.myanimelist.net/images/voiceactors...      Ilka Pinheiro   

      given_name family_name                   birthday  favorites  \
4171         ナツコ          高橋                        NaN         19   
4172         ナツコ          高橋                        NaN         19   
5202         NaN         NaN                        NaN         11   
6018          敏夫          鈴木  1948-08-19T00:00:00+00:00        354   
6019          敏夫          鈴木  1948-08-19T00:00:00+00:00        354   
...          ...         ...                        ...        ...   
64387         絵奈          西川                        NaN          0   
64402        NaN         NaN  1998-07-05T00:00:00+00:00          0   
64412        NaN         NaN                        NaN          0   
64414        NaN         NaN                        NaN          0   
64417        NaN         NaN                        NaN          0   

             relevant_location  
4171             Paris, France  
4172   Cape Town, South Africa  
5202           Yokohama, Japan  
6018             Paris, France  
6019             Mumbai, India  
...                        ...  
64387            Paris, France  
64402        São Paulo, Brazil  
64412          Berlin, Germany  
64414            Mumbai, India  
64417            New York, USA  

[105 rows x 10 columns]

In [49]:
person_details_df[person_details_df.isna().any(axis=1)]


person_mal_id                                                url  \
0                  1     https://myanimelist.net/people/1/Tomokazu_Seki   
2                  3    https://myanimelist.net/people/3/Satsuki_Yukino   
9                 10  https://myanimelist.net/people/10/Johnny_Yong_...   
10                11  https://myanimelist.net/people/11/Kouichi_Yama...   
11                12      https://myanimelist.net/people/12/Steven_Blum   
...              ...                                                ...   
76694          90011  https://myanimelist.net/people/90011/Nanako_Ki...   
76695          90012         https://myanimelist.net/people/90012/Pamon   
76696          90013  https://myanimelist.net/people/90013/Tomoru_Emoto   
76697          90014        https://myanimelist.net/people/90014/Hirari   
76698          90015  https://myanimelist.net/people/90015/Daisuke_T...   

                          website_url  \
0                                 NaN   
2                                 NaN   
9        https://johnnyyongbosch.com/   
10                                NaN   
11     http://www.steveblumvoices.com   
...                               ...   
76694                             NaN   
76695                             NaN   
76696                             NaN   
76697          https://hirari.2-d.jp/   
76698                             NaN   

                                               image_url               name  \
0      https://cdn.myanimelist.net/images/voiceactors...      Tomokazu Seki   
2      https://cdn.myanimelist.net/images/voiceactors...     Satsuki Yukino   
9      https://cdn.myanimelist.net/images/voiceactors...  Johnny Yong Bosch   
10     https://cdn.myanimelist.net/images/voiceactors...   Kouichi Yamadera   
11     https://cdn.myanimelist.net/images/voiceactors...        Steven Blum   
...                                                  ...                ...   
76694  https://cdn.myanimelist.net/img/sp/icon/apple-...   Nanako Kishimoto   
76695  https://cdn.myanimelist.net/img/sp/icon/apple-...              Pamon   
76696  https://cdn.myanimelist.net/img/sp/icon/apple-...       Tomoru Emoto   
76697  https://cdn.myanimelist.net/img/sp/icon/apple-...             Hirari   
76698  https://cdn.myanimelist.net/img/sp/icon/apple-...  Daisuke Taniguchi   

      given_name family_name                   birthday  favorites  \
0             智一           関  1972-09-08T00:00:00+00:00       6219   
2            さつき         ゆきの  1970-05-25T00:00:00+00:00       1777   
9            NaN         NaN  1976-01-06T00:00:00+00:00      12869   
10            宏一          山寺  1961-06-17T00:00:00+00:00       1574   
11           NaN         NaN  1960-04-29T00:00:00+00:00       7749   
...          ...         ...                        ...        ...   
76694         七子          岸本                        NaN          0   
76695        NaN          파몬                        NaN          0   
76696        ともる          柄本                        NaN          0   
76697        NaN         ひらり                        NaN          0   
76698         大介          谷口                        NaN          0   

             relevant_location  
0              Berlin, Germany  
2                Madrid, Spain  
9             Los Angeles, USA  
10               New York, USA  
11           São Paulo, Brazil  
...                        ...  
76694            Mumbai, India  
76695             Tokyo, Japan  
76696             Tokyo, Japan  
76697            Paris, France  
76698  Cape Town, South Africa  

[72168 rows x 10 columns]

In [50]:
person_details_df.dtypes
#we need to change birthday from object to data

person_mal_id         int64
url                  object
website_url          object
image_url            object
name                 object
given_name           object
family_name          object
birthday             object
favorites             int64
relevant_location    object
dtype: object

In [51]:
person_details_df.isna().sum()
#we have to check the nan values

person_mal_id            0
url                      0
website_url          59264
image_url                0
name                     2
given_name           30253
family_name          18641
birthday             59461
favorites                0
relevant_location        0
dtype: int64

We can join the two tables person_details_df and person_alternate_names_df having the keys that match.
Putting the alternate names in a new column called alt_name and having a list of those inside

### SIXTH DATASET

In [52]:
person_anime_works_df

person_mal_id                     position  anime_mal_id
0                   1       Theme Song Performance          3080
1                   1    Inserted Song Performance         15699
2                   1  Theme Song Performance (OP)           247
3                   1       Theme Song Performance           258
4                   1  Theme Song Performance (ED)         34825
...               ...                          ...           ...
458086          89951         In-Between Animation         11001
458087          89951                Key Animation         55092
458088          89951     Key Animation (eps 4, 9)         20053
458089          89951                Key Animation         50553
458090          89965          Music Not Yet Aired         62579

[458091 rows x 3 columns]

In [53]:
person_anime_works_df.dtypes
#the types are correct

person_mal_id     int64
position         object
anime_mal_id      int64
dtype: object

In [54]:
person_anime_works_df.isna().sum()
#There's no nan value

person_mal_id    0
position         0
anime_mal_id     0
dtype: int64

### SEVENTH DATASET

In [55]:
stats_df

mal_id  watching  completed  on_hold  dropped  plan_to_watch   total  \
0       59356         7        146        4       20             20     197   
1       56036        21        770        8       29            113     941   
2        2928       451      14953      302      349           6472   22527   
3        3269       726      22790      452      537           9762   34267   
4        4469       241       6918      182      266           3528   11135   
...       ...       ...        ...      ...      ...            ...     ...   
28950   59421     11349      17353      574     1440          15729   46445   
28951   31245     10332     140676     2989     1416          94249  249662   
28952   36305       928      16119      388      243           9656   27334   
28953   34895       896       2387      375      356           1512    5526   
28954   56606         4         83        2       15             20     124   

       score_1_votes  score_1_percentage  score_2_votes  ...  score_6_votes  \
0                2.0                 2.2            0.0  ...           33.0   
1                5.0                 1.0            6.0  ...          138.0   
2              101.0                 1.0           93.0  ...         2054.0   
3              120.0                 0.8          156.0  ...         2457.0   
4               83.0                 1.9          104.0  ...          888.0   
...              ...                 ...            ...  ...            ...   
28950           54.0                 0.3           40.0  ...         2303.0   
28951          357.0                 0.3          484.0  ...        15972.0   
28952           48.0                 0.5           28.0  ...         1839.0   
28953          116.0                 6.1           80.0  ...          437.0   
28954            3.0                 5.2            1.0  ...           15.0   

       score_6_percentage  score_7_votes  score_7_percentage  score_8_votes  \
0                    36.3           19.0                20.9            2.0   
1                    27.4          144.0                28.6           81.0   
2                    21.1         2709.0                27.8         1500.0   
3                    16.0         4157.0                27.0         3075.0   
4                    20.6          871.0                20.2          592.0   
...                   ...            ...                 ...            ...   
28950                14.5         5545.0                34.9         3919.0   
28951                15.3        34370.0                33.0        24387.0   
28952                18.3         3258.0                32.5         2009.0   
28953                23.2          280.0                14.8          170.0   
28954                25.9            7.0                12.1            1.0   

       score_8_percentage  score_9_votes  score_9_percentage  score_10_votes  \
0                     2.2            0.0                 0.0             1.0   
1                    16.1           17.0                 3.4            40.0   
2                    15.4          875.0                 9.0           608.0   
3                    20.0         1919.0                12.5          1400.0   
4                    13.7          308.0                 7.1           315.0   
...                   ...            ...                 ...             ...   
28950                24.7         1598.0                10.1          1235.0   
28951                23.4        10480.0                10.1          7589.0   
28952                20.0          887.0                 8.8           936.0   
28953                 9.0           68.0                 3.6           125.0   
28954                 1.7            0.0                 0.0             4.0   

       score_10_percentage  
0                      1.1  
1                      8.0  
2                      6.2  
3                      9.1  
4                      7.3  
...                    ..

In [56]:
stats_df.dtypes
#we can swap *_votes to int instead of floats

mal_id                   int64
watching                 int64
completed                int64
on_hold                  int64
dropped                  int64
plan_to_watch            int64
total                    int64
score_1_votes          float64
score_1_percentage     float64
score_2_votes          float64
score_2_percentage     float64
score_3_votes          float64
score_3_percentage     float64
score_4_votes          float64
score_4_percentage     float64
score_5_votes          float64
score_5_percentage     float64
score_6_votes          float64
score_6_percentage     float64
score_7_votes          float64
score_7_percentage     float64
score_8_votes          float64
score_8_percentage     float64
score_9_votes          float64
score_9_percentage     float64
score_10_votes         float64
score_10_percentage    float64
dtype: object

In [57]:
stats_df.isna().sum()
#there are 430 series without any vote (?)
#we need to understand which ones are they

mal_id                   0
watching                 0
completed                0
on_hold                  0
dropped                  0
plan_to_watch            0
total                    0
score_1_votes          430
score_1_percentage     430
score_2_votes          430
score_2_percentage     430
score_3_votes          430
score_3_percentage     430
score_4_votes          430
score_4_percentage     430
score_5_votes          430
score_5_percentage     430
score_6_votes          430
score_6_percentage     430
score_7_votes          430
score_7_percentage     430
score_8_votes          430
score_8_percentage     430
score_9_votes          430
score_9_percentage     430
score_10_votes         430
score_10_percentage    430
dtype: int64

In [58]:
stats_df[stats_df['score_1_votes'].isna()]


mal_id  watching  completed  on_hold  dropped  plan_to_watch  total  \
124     62496         6          0        2        0            652    660   
134     61637         0          0        0        0           2128   2128   
239     62002         1          0        0        0           1504   1505   
285     61925         3          0        1        1          11270  11275   
437     60917         0          0        0        0            739    739   
...       ...       ...        ...      ...      ...            ...    ...   
28835   61947         0          0        0        0             59     59   
28859   61646         0          0        0        0            120    120   
28860   61640         0          0        0        0            131    131   
28898   57990         0          0        0        0           2308   2308   
28903   50159         3          0        2        0          33878  33883   

       score_1_votes  score_1_percentage  score_2_votes  ...  score_6_votes  \
124              NaN                 NaN            NaN  ...            NaN   
134              NaN                 NaN            NaN  ...            NaN   
239              NaN                 NaN            NaN  ...            NaN   
285              NaN                 NaN            NaN  ...            NaN   
437              NaN                 NaN            NaN  ...            NaN   
...              ...                 ...            ...  ...            ...   
28835            NaN                 NaN            NaN  ...            NaN   
28859            NaN                 NaN            NaN  ...            NaN   
28860            NaN                 NaN            NaN  ...            NaN   
28898            NaN                 NaN            NaN  ...            NaN   
28903            NaN                 NaN            NaN  ...            NaN   

       score_6_percentage  score_7_votes  score_7_percentage  score_8_votes  \
124                   NaN            NaN                 NaN            NaN   
134                   NaN            NaN                 NaN            NaN   
239                   NaN            NaN                 NaN            NaN   
285                   NaN            NaN                 NaN            NaN   
437                   NaN            NaN                 NaN            NaN   
...                   ...            ...                 ...            ...   
28835                 NaN            NaN                 NaN            NaN   
28859                 NaN            NaN                 NaN            NaN   
28860                 NaN            NaN                 NaN            NaN   
28898                 NaN            NaN                 NaN            NaN   
28903                 NaN            NaN                 NaN            NaN   

       score_8_percentage  score_9_votes  score_9_percentage  score_10_votes  \
124                   NaN            NaN                 NaN             NaN   
134                   NaN            NaN                 NaN             NaN   
239                   NaN            NaN                 NaN             NaN   
285                   NaN            NaN                 NaN             NaN   
437                   NaN            NaN                 NaN             NaN   
...                   ...            ...                 ...             ...   
28835                 NaN            NaN                 NaN             NaN   
28859                 NaN            NaN                 NaN             NaN   
28860                 NaN            NaN                 NaN             NaN   
28898                 NaN            NaN                 NaN             NaN   
28903                 NaN            NaN                 NaN             NaN   

       score_10_percentage  
124                    NaN  
134                    NaN  
239                    NaN  
285                    NaN  
437                    NaN  
...                    ...  
28835   

In [59]:
details_df.query('mal_id == 61558')

mal_id                                          title  \
529   61558  Aikatsu! x PriPara the Movie: Deai no Kiseki!   

                      title_japanese  \
529  アイカツ！×プリパラ THE MOVIE -出会いのキセキ！-   

                                                   url  \
529  https://myanimelist.net/anime/61558/Aikatsu_x_...   

                                             image_url   type         status  \
529  https://cdn.myanimelist.net/images/anime/1723/...  Movie  Not yet aired   

     score  scored_by                 start_date  ... demographics    source  \
529    NaN        NaN  2025-10-10T00:00:00+00:00  ...           []  Original   

     rating  episodes  season  year producers explicit_genres licensors  \
529     NaN       1.0     NaN   NaN        []              []        []   

    streaming  
529        []  

[1 rows x 29 columns]

### EIGHTH DATASET 

In [60]:
ratings_df

username  anime_id         status  score  is_rewatching  \
0          --------788     30276       watching      7            0.0   
1          --------788     28851      completed      7            0.0   
2          --------788     41168      completed      7            0.0   
3          --------788     22199      completed     10            0.0   
4          --------788     16498      completed     10            0.0   
...                ...       ...            ...    ...            ...   
124298352      arizkim     52305  plan_to_watch      0            0.0   
124298353      arizkim      4224  plan_to_watch      0            0.0   
124298354      arizkim     54790  plan_to_watch      0            0.0   
124298355      arizkim     53835  plan_to_watch      0            0.0   
124298356      arizkim     33352  plan_to_watch      0            0.0   

           num_watched_episodes  
0                             3  
1                             1  
2                             1  
3                            24  
4                            25  
...                         ...  
124298352                     0  
124298353                     0  
124298354                     0  
124298355                     0  
124298356                     0  

[124298357 rows x 6 columns]

In [61]:
# we have to understand the sense of "num_watched_episodes" and the link with "is_rewatching"
ratings_df.query('is_rewatching == 1')

username  anime_id     status  score  is_rewatching  \
934                   CKK2         6  completed      9            1.0   
8939       FollowYourHeart      1210  completed     10            1.0   
11047         ----Haku----     39017  completed      7            1.0   
22879            KarioBaka     10087  completed      9            1.0   
22928            KarioBaka     14467  completed      0            1.0   
...                    ...       ...        ...    ...            ...   
124291819      arissabelle      2167  completed      0            1.0   
124291878      arissabelle     33352  completed     10            1.0   
124291883      arissabelle      6547    on_hold      0            1.0   
124297299      ariyanroy04     46569  completed      0            1.0   
124297323      ariyanroy04        19  completed      0            1.0   

           num_watched_episodes  
934                           8  
8939                         24  
11047                         3  
22879                         3  
22928                         6  
...                         ...  
124291819                     0  
124291878                     0  
124291883                     1  
124297299                    13  
124297323                    28  

[103341 rows x 6 columns]

In [62]:
details_df.query('mal_id == 918')

mal_id    title title_japanese  \
7358     918  Gintama             銀魂   

                                            url  \
7358  https://myanimelist.net/anime/918/Gintama   

                                              image_url type           status  \
7358  https://cdn.myanimelist.net/images/anime/10/73...   TV  Finished Airing   

      score  scored_by                 start_date  ... demographics source  \
7358   8.93   415923.0  2006-04-04T00:00:00+00:00  ...  ['Shounen']  Manga   

                         rating  episodes  season    year  \
7358  PG-13 - Teens 13 or older     201.0  spring  2006.0   

                                              producers explicit_genres  \
7358  ['TV Tokyo', 'Aniplex', 'Dentsu', 'Trinity Sou...              []   

                                licensors                   streaming  
7358  ['Sentai Filmworks', 'Crunchyroll']  ['Crunchyroll', 'Netflix']  

[1 rows x 29 columns]

In [63]:
# change "is_rewatching" from float to int or boolean
ratings_df.dtypes

username                 object
anime_id                  int64
status                   object
score                     int64
is_rewatching           float64
num_watched_episodes      int64
dtype: object

In [31]:
ratings_df["is_rewatching"] = ratings_df["is_rewatching"].astype("Int64")

NameError: name 'ratings_df' is not defined

In [64]:
ratings_df.loc[ratings_df.duplicated(subset=['username', 'anime_id'])]

username  anime_id    status  score  is_rewatching  \
44469797    W00F1234     59457   dropped      3            0.0   
74614909   Pavle2009     44511  watching      0            0.0   
74616569  Doopelsdoo     59459  watching      0            0.0   
74626628    Door_mp3     59644  watching      8            0.0   
74626629    Door_mp3     60564  watching      8            0.0   
74626630    Door_mp3     61159  watching      9            0.0   

          num_watched_episodes  
44469797                     1  
74614909                     7  
74616569                     9  
74626628                     6  
74626629                     5  
74626630                     5

In [65]:
# we check for Nan values. 
# if it necessary check if the num_watched_episodes is greater than number of episodes of anime, we can remove the Nan values and put one or zero. 
ratings_df.isna().sum()

username                      7
anime_id                      0
status                        0
score                         0
is_rewatching           3797321
num_watched_episodes          0
dtype: int64

In [66]:
# check the rows with "Nan" values
ratings_df[ratings_df['username'].isna()]

username  anime_id         status  score  is_rewatching  \
100051758      NaN      1482       watching      9            0.0   
100051759      NaN      1735       watching      9            0.0   
100051760      NaN       121      completed     10            0.0   
100051761      NaN       136      completed      8            0.0   
100051762      NaN       269        on_hold      7            0.0   
100051763      NaN      1818        on_hold      7            0.0   
100051764      NaN      1535  plan_to_watch      6            0.0   

           num_watched_episodes  
100051758                    35  
100051759                    21  
100051760                    51  
100051761                    62  
100051762                    33  
100051763                    11  
100051764                     4

Check this username that there is in the profiles_df

In [32]:
# we want to see all duplicates to understand if we have to drop or not
ratings_df.loc[ratings_df.duplicated(subset=['username', 'anime_id', 'status'], keep=False)]

NameError: name 'ratings_df' is not defined

### NINTH DATASET

In [30]:
characters_df

character_mal_id                                                url  \
0               280386.0  https://myanimelist.net/character/280386/Envi_...   
1               280354.0    https://myanimelist.net/character/280354/Eleven   
2               280353.0      https://myanimelist.net/character/280353/Stud   
3               280352.0     https://myanimelist.net/character/280352/Judge   
4               280339.0  https://myanimelist.net/character/280339/Eiji_...   
...                  ...                                                ...   
209958          282276.0  https://myanimelist.net/character/282276/Farra...   
209959          282277.0  https://myanimelist.net/character/282277/Harri...   
209960          282278.0       https://myanimelist.net/character/282278/Rob   
209961          282281.0     https://myanimelist.net/character/282281/Grimm   
209962          282284.0  https://myanimelist.net/character/282284/Joel_...   

                      name       name_kanji  \
0       Envi Mel Champagne  エンヴィ・メル・シャンパーニュ   
1                   Eleven             イレヴン   
2                     Stud             スタッド   
3                    Judge             ジャッジ   
4            Eiji Kurokawa            黒川 英治   
...                    ...              ...   
209958  Farrah Van Dorothy     ファラ・ヴァン・ドロシー   
209959         Harris Mead          ハリス・ミード   
209960                 Rob               ロブ   
209961               Grimm              グリム   
209962       Joel Widdison             ジョエル   

                                                    image  favorites about  
0       https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
1       https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
2       https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
3       https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
4       https://cdn.myanimelist.net/img/sp/icon/apple-...        0.0   NaN  
...                                                   ...        ...   ...  
209958  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
209959  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
209960  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
209961  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
209962  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  

[209963 rows x 7 columns]

In [68]:
# check types of dataset columns
characters_df.dtypes

character_mal_id    float64
url                  object
name                 object
name_kanji           object
image                object
favorites           float64
about                object
dtype: object

In [32]:
# change "character_mal_id" and "favorites" from float to int
characters_df["character_mal_id"] = characters_df["character_mal_id"].astype("Int64")
characters_df["favorites"] = characters_df["favorites"].astype("Int64")

In [70]:
characters_df.describe()

character_mal_id    favorites
count          209961.0     209961.0
mean      150579.010283    57.616138
std        85243.624423  1197.586488
min                 1.0          0.0
25%             74581.0          0.0
50%            160191.0          0.0
75%            225373.0          2.0
max            282284.0     175632.0

In [71]:
# we have only 2 rows where all columns are Nan, the rows with Nan values in "name_kanji" and "about" we shouldn't drop because they have other values that are important.
characters_df.isna().sum()

character_mal_id        2
url                     2
name                    2
name_kanji          55480
image                   2
favorites               2
about               96976
dtype: int64

In [30]:
# here we want to check if the Nan values are concentrate in only two rows
characters_df[characters_df['character_mal_id'].isna()]

character_mal_id  url name name_kanji image  favorites about
209733               NaN  NaN  NaN        NaN   NaN        NaN   NaN
209734               NaN  NaN  NaN        NaN   NaN        NaN   NaN

In [23]:
# Apart "name_kanji" and "about" the others Nan values are concentrate in two rows so we drop the two rows with all columns Nan
characters_df = characters_df.dropna(how='all')

In [33]:
# we want to see all duplicates to understand if we have to drop or not
characters_df.loc[characters_df.duplicated(subset=['character_mal_id', 'url', 'name'], keep=False)]

character_mal_id                                                url  \
853               279200  https://myanimelist.net/character/279200/Shing...   
854               279200  https://myanimelist.net/character/279200/Shing...   
909               279142      https://myanimelist.net/character/279142/Kaho   
910               279142      https://myanimelist.net/character/279142/Kaho   
1166              278865  https://myanimelist.net/character/278865/Yoshi...   
...                  ...                                                ...   
209789            282042     https://myanimelist.net/character/282042/Alina   
209849            282115    https://myanimelist.net/character/282115/Cavitt   
209850            282115    https://myanimelist.net/character/282115/Cavitt   
209858            282131  https://myanimelist.net/character/282131/Lieut...   
209859            282131  https://myanimelist.net/character/282131/Lieut...   

                     name name_kanji  \
853      Shingo Shimazaki      島崎 真悟   
854      Shingo Shimazaki      島崎 真悟   
909                  Kaho        NaN   
910                  Kaho        NaN   
1166    Yoshirou Sonozaki      園崎 義郎   
...                   ...        ...   
209789              Alina        アリナ   
209849             Cavitt      キャビット   
209850             Cavitt      キャビット   
209858  Lieutenant Rencon     レンコン中尉   
209859  Lieutenant Rencon     レンコン中尉   

                                                    image  favorites  \
853     https://cdn.myanimelist.net/images/characters/...          1   
854     https://cdn.myanimelist.net/images/characters/...          1   
909     https://cdn.myanimelist.net/images/characters/...          0   
910     https://cdn.myanimelist.net/images/characters/...          0   
1166    https://cdn.myanimelist.net/images/characters/...          0   
...                                                   ...        ...   
209789  https://cdn.myanimelist.net/images/characters/...          0   
209849  https://cdn.myanimelist.net/images/characters/...          0   
209850  https://cdn.myanimelist.net/images/characters/...          0   
209858  https://cdn.myanimelist.net/images/characters/...          0   
209859  https://cdn.myanimelist.net/images/characters/...          0   

                                                    about  
853                                                   NaN  
854                                                   NaN  
909                                                   NaN  
910                                                   NaN  
1166              The owner of the restaurant Angel Mort.  
...                                                   ...  
209789                                                NaN  
209849  Lieutenant Rencon's younger sister.\n\nNo voic...  
209850  Lieutenant Rencon's younger sister.\n\nNo voic...  
209858  Former lieutenant of Mar Expedition and older ...  
209859  Former lieutenant of Mar Expedition and older ...  

[796 rows x 7 columns]

In [25]:
# we drop the duplicates because they have all same values 
characters_df = characters_df.drop_duplicates(subset=['character_mal_id', 'url', 'name'], keep='first')

### TENTH DATASET

In [76]:
# role of character anime
character_anime_works_df

anime_mal_id  character_mal_id    character_name        role
0               2928              5781             Atoli        Main
1               2928                33             Haseo        Main
2               2928                32              Ovan        Main
3               2928                34             Shino        Main
4               2928              5785              Aina  Supporting
...              ...               ...               ...         ...
236811         31245            137157    Shibasaki, Ken  Supporting
236812         36305            136064  Hamanaka, Midori        Main
236813         36305            133916      Narumi, Sena        Main
236814         36305            124942   Hayasaka, Akari  Supporting
236815         34895            155887  Shiraha, Setsuna        Main

[236816 rows x 4 columns]

In [8]:
# check types of columns
character_anime_works_df.dtypes

anime_mal_id         int64
character_mal_id     int64
character_name      object
role                object
dtype: object

In [9]:
# check the number of Nan value
character_anime_works_df.isna().sum()

anime_mal_id        0
character_mal_id    0
character_name      0
role                0
dtype: int64

In [79]:
# check the number of duplicates
character_anime_works_df.loc[character_anime_works_df.duplicated(subset=['anime_mal_id', 'character_mal_id'])]

Empty DataFrame
Columns: [anime_mal_id, character_mal_id, character_name, role]
Index: []

There is no need to clean this dataset, it haven't no Nan values or duplicates

### ELEVENTH DATASET

In [40]:
person_voice_works_df

person_mal_id        role  anime_mal_id  character_mal_id  language
0                   1        Main         55830              2514  Japanese
1                   1  Supporting         60602              2822  Japanese
2                   1  Supporting         59229            140499  Japanese
3                   1  Supporting         60427            275856  Japanese
4                   1  Supporting         62067            190335  Japanese
...               ...         ...           ...               ...       ...
489511          89839  Supporting         40111            274622  Mandarin
489512          89840  Supporting         60544            266412  Mandarin
489513          89841  Supporting         60544            266416  Mandarin
489514          89842  Supporting         36896            279922  Japanese
489515          89952  Supporting         60260            279778  Japanese

[489516 rows x 5 columns]

In [10]:
person_voice_works_df.dtypes

person_mal_id        int64
role                object
anime_mal_id         int64
character_mal_id     int64
language            object
dtype: object

In [11]:
person_voice_works_df.isna().sum()

person_mal_id       0
role                0
anime_mal_id        0
character_mal_id    0
language            0
dtype: int64

In [41]:
# check if the duplicates are in all columns
person_voice_works_df.loc[person_voice_works_df.duplicated(keep=False)]

person_mal_id        role  anime_mal_id  character_mal_id  \
173909           5742        Main          1011               668   
173910           5742        Main          1008               668   
173911           5742        Main          1010               668   
173912           5742        Main          1007               668   
173913           5742        Main           792               668   
...               ...         ...           ...               ...   
478139          77591  Supporting         41025            151299   
478140          77591        Main           431               508   
478141          77591        Main          2202              7952   
478142          77591  Supporting          1069             77729   
478143          77591  Supporting          2225             30307   

               language  
173909          English  
173910          English  
173911          English  
173912          English  
173913          English  
...                 ...  
478139  Portuguese (BR)  
478140  Portuguese (BR)  
478141  Portuguese (BR)  
478142  Portuguese (BR)  
478143  Portuguese (BR)  

[1488 rows x 5 columns]

In [36]:
# drop the duplicates because they have all same values
person_voice_works_df = person_voice_works_df.drop_duplicates(keep='first')

### TWELFTH DATASET

In [86]:
# Should we delete the last five columns
profiles_df

username  gender      birthday       location        joined  \
0         ishikawas     NaN           NaN    South Korea           NaN   
1              CKK2     NaN           NaN  United States   Dec 1, 2018   
2       --------788  Female           NaN         Mexico   Oct 4, 2022   
3        potatoaris     NaN           NaN          Spain   Oct 2, 2018   
4         Rinrintan     NaN           NaN          Japan  May 12, 2019   
...             ...     ...           ...            ...           ...   
337150  ariyanroy04     NaN           NaN  United States  Mar 30, 2021   
337151   ariyanvk18    Male           NaN         Turkey  Jan 24, 2022   
337152     ariyoskz     NaN           NaN        Germany   Jul 7, 2022   
337153    arizima23  Female  Jun 23, 1998          Spain  Mar 15, 2019   
337154      arizkim     NaN           NaN   South Africa   Dec 7, 2020   

       watching completed on_hold dropped plan_to_watch  
0           NaN       NaN     NaN     NaN           NaN  
1             3       182      15       0           405  
2             1        64       0       0             1  
3             5         1       0       0             4  
4            20       311      40      16            34  
...         ...       ...     ...     ...           ...  
337150       35       130       0       0           489  
337151        7       101       1      12            21  
337152        1       155       2      11            15  
337153       15        64       0       4             4  
337154        7        38       1       0            33  

[337155 rows x 10 columns]

In [26]:
# check if the types are right for each field
profiles_df.dtypes

username                 object
gender                   object
birthday         datetime64[ns]
location                 object
joined           datetime64[ns]
watching                 object
completed                object
on_hold                  object
dropped                  object
plan_to_watch            object
dtype: object

In [25]:
# change types of columns "birthday" and "joined" from object to date and the others columns that they should be int
profiles_df["birthday"] = pd.to_datetime(profiles_df["birthday"], errors='coerce')
profiles_df["joined"] = pd.to_datetime(profiles_df["joined"], errors='coerce')

In [23]:
# check the number of Nan in each column
profiles_df.isna().sum()

username              1
gender           170876
birthday         251156
location              0
joined             1676
watching           1678
completed          1678
on_hold            1678
dropped            1678
plan_to_watch      1678
dtype: int64

Even if there are many "Nan" values we can't drop any rows, one example is for the only username Nan, we can't drop it because if we drop it with this we delete seven others values we need as we can see in the cell below 

In [91]:
profiles_df.query('username.isna()')

username gender birthday location        joined watching completed  \
271344      NaN    NaN      NaN   Mexico  Jul 28, 2007        2         2   

       on_hold dropped plan_to_watch  
271344       2       0             1

In [92]:
ratings_df.query('username.isna()')

username  anime_id         status  score  is_rewatching  \
100051758      NaN      1482       watching      9            0.0   
100051759      NaN      1735       watching      9            0.0   
100051760      NaN       121      completed     10            0.0   
100051761      NaN       136      completed      8            0.0   
100051762      NaN       269        on_hold      7            0.0   
100051763      NaN      1818        on_hold      7            0.0   
100051764      NaN      1535  plan_to_watch      6            0.0   

           num_watched_episodes  
100051758                    35  
100051759                    21  
100051760                    51  
100051761                    62  
100051762                    33  
100051763                    11  
100051764                     4

In [24]:
# check if there is any duplicate on "username"
profiles_df.loc[profiles_df.duplicated(subset=['username'], keep='first')]

Empty DataFrame
Columns: [username, gender, birthday, location, joined, watching, completed, on_hold, dropped, plan_to_watch]
Index: []

### THIRTEENTH DATASET

In [94]:
recommendations_df

mal_id  recommendation_mal_id
0         3269                    317
1         3269                   6922
2         3269                    299
3         3269                   3446
4         3269                   5681
...        ...                    ...
105244   31245                   1689
105245   31245                  35434
105246   31245                  31798
105247   31245                  21995
105248   36305                  42897

[105249 rows x 2 columns]

In [27]:
recommendations_df.dtypes

mal_id                   int64
recommendation_mal_id    int64
dtype: object

In [28]:
recommendations_df.isna().sum()

mal_id                   0
recommendation_mal_id    0
dtype: int64

In [29]:
recommendations_df.loc[profiles_df.duplicated(keep='first')]

Empty DataFrame
Columns: [mal_id, recommendation_mal_id]
Index: []